content/number-theory/MillerRabin.h

```diff
@@ -1,30 +1,27 @@
 /**
- * Author: Lukas Polacek
- * Date: 2010-01-26
+ * Author: chilli, c1729
+ * Date: 2019-03-28
  * License: CC0
- * Source: TopCoder tutorial
- * Description: Miller-Rabin primality probabilistic test.
- * Probability of failing one iteration is at most 1/4. 15 iterations should be
- * enough for 50-bit numbers.
- * Time: 15 times the complexity of $a^b \mod c$.
+ * Description: Deterministic Miller-Rabin primality test.
+ * Guaranteed to be correct, 2x as probabilistic version for <1e9, 3x as fast
+ * for 1e15
+ * Time: 7 times the complexity of $a^b \mod c$.
  */
 #pragma once
 
 #include "ModMulLL.h"
 
-bool prime(ull p) {
-	if (p == 2) return true;
-	if (p == 1 || p % 2 == 0) return false;
-	ull s = p - 1;
-	while (s % 2 == 0) s /= 2;
-	rep(i,0,15) {
-		ull a = rand() % (p - 1) + 1, tmp = s;
-		ull mod = mod_pow(a, tmp, p);
-		while (tmp != p - 1 && mod != 1 && mod != p - 1) {
-			mod = mod_mul(mod, mod, p);
-			tmp *= 2;
-		}
-		if (mod != p - 1 && tmp % 2 == 0) return false;
-	}
-	return true;
+bool prime(ull n) {
+    vector<ull> ps({2, 3, 5, 13, 19, 73, 193, 407521, 299210837});
+    vector<ull> cs({2, 325, 9375, 28178, 450775, 9780504, 1795265022});
+    if (n <= 1 || any_of(all(ps), [&](ull p) { return n % p == 0; }))
+        return count(all(ps), n) > 0;
+    ull d = n - 1, s = 0;
+    while (!(d & 1)) d >>= 1, s++;
+    return !any_of(all(cs), [&](ull a) {
```

Comment:
The `any_of` involves quite a bit of boilerplate and perhaps overhead if the compiler fails to inline it. How about
```c++
for (ull a : {2, 325, 9375, 28178, 450775, 9780504, 1795265022})
  for (ull i = 0; ...) ...
return true;

Reply:
The semantics here are a bit annoying to duplicate with regular loops - largely because you need to break out of two nested loops.

Here's a couple of other options I came up with, in varying levels of jankness/seriousness

```cpp
for (ull a : {2, 325, 9375, 28178, 450775, 9780504, 1795265022}) {
    for (ull i = 0, p = binExp(a, d, n); i < s; i++, p = binExp(p, 2, n))
        if (p == n - 1 || (i == 0 && p == 1))
            goto end;
    return false;
end:;
}
```

```cpp
for (ull a : {2, 325, 9375, 28178, 450775, 9780504, 1795265022}) {
    ull p = binExp(a, d, n), i = 0;
    if (p == 1)
        continue;
    for (; p != n - 1 && i < s; i++)
        p = binExp(p, 2, n);
    if (i == s)
        return false;
}
return true;
```

```cpp
for (ull a : {2, 325, 9375, 28178, 450775, 9780504, 1795265022}) {
    if ([&] {
            for (ull i = 0, p = binExp(a, d, n); i < s; i++, p = binExp(p, 2, n))
                if (p == n - 1 || (i == 0 && p == 1))
                    return false;
            return true;
        }())
        return false;
}
return true;
```
```cpp
for (ull a : {2, 325, 9375, 28178, 450775, 9780504, 1795265022}) {
    bool found = false;
    for (ull i = 0, p = binExp(a, d, n); i < s; i++, p = binExp(p, 2, n))
        if (p == n - 1 || (i == 0 && p == 1)) {
            found = true;
            break;
        }
    if (!found)
        return false;
}
return true;
```

Reply:
I like the last one the most

Reply:
Ah, I see. Agreed, the last one looks good. Curiously, it's very similar to KACTL's current version, except that the last condition is `if (i == n - 1 && i > d) return false;` rather than `if (mod != p - 1 && tmp % 2 == 0) return false;`. I don't know what that difference means...

Reply:
Thinking about it more... I think what I have right now has a slight bug. Gimme a bit to fix it.

Reply:
The conditions that are equivalent (and this isn't obvious, it relies on specific knowledge about Miller-Rabin) are `i==n-1 && p!=n-1` or `p!=n-1 && i!=d` (the second one is more directly equivalent to the current KACTL condition).

I'll stick with the current KACTL condition - it's shorter :^)